### PR TITLE
Update vector processing

### PR DESCRIPTION
**Pull request by Hdt80bro**
Adds metamethods for vectors (quaternions in particular) and cleans up lots of vector code.

### CODE REVIEW

File: lua/system/repr.lua [hunk 1]
```diff
@@ -22,8 +22,10 @@ local function IsState(t)
     return (t.__State and true) or false
 end
 
+-- Vector2, Vector, and Quaternion all share the same metatable
+local vector_metatable = getmetatable(Vector2(0, 0))
 local function IsVector(t)
-    return (t.x and t.y and t.z and true) or false 
+    return getmetatable(t) == vector_metatable
```
Comment (Garanas) [inline lua/system/repr.lua:28]: This is quite smart 😄 !

Comment (Hdt80bro) [reply lua/system/repr.lua:28]: It also lets it detect `Vector2`s and `Quaternion`s :)

**Comment by Garanas** [comment]
The orientation of units is off

image

**Comment by Hdt80bro** [comment]
Ugg, I thought I finally figured it out.

**Comment by Garanas** [comment]
@lL1l1 would you be interested in finalizing this pull request? Having proper functions would make it easier to read, understand and maintain

**Comment by lL1l1** [comment]
I'll work on it after I finish up my 3 draft PRs, I think it will be very useful but I don't want to pile up stuff.

**Comment by MrRowey** [comment]
@Hdt80bro or @lL1l1 Any update on this ?

**Comment by Garanas** [comment]
Closed in favor of https://github.com/FAForever/fa/pull/6438

**Pull request closed without merging**